lib/decoder.js

```diff
@@ -1,436 +1,326 @@
-      case 0xd2:
-        return 5
-      case 0xd3:
-        return 9
-      case 0xd4:
-        return 3
-      case 0xd5:
-        return 4
-      case 0xd6:
-        return 6
-      case 0xd7:
-        return 10
-      case 0xd8:
-        return 18
-      case 0xd9:
-        return 2
-      case 0xda:
-        return 3
-      case 0xdb:
-        return 5
-      case 0xde:
-        return 3
-      default:
-        return -1
-    }
-  }
-
-  function hasMinBufferSize (first, length) {
-    var size = getSize(first)
-
-    if (size !== -1 && length < size) {
-      return false
-    } else {
-      return true
-    }
-  }
-
   function isValidDataSize (dataLength, bufLength, headerLength) {
     return bufLength >= headerLength + dataLength
   }
 
-  function buildDecodeResult (value, bytesConsumed) {
-    return {
-      value: value,
-      bytesConsumed: bytesConsumed
-    }
-  }
-
   function decode (buf) {
+    // TODO: Make it into ensureBl handler ?
     if (!(buf instanceof bl)) {
       buf = bl().append(buf)
     }
 
     var result = tryDecode(buf)
-    if (result) {
-      buf.consume(result.bytesConsumed)
-      return result.value
-    } else {
-      throw new IncompleteBufferError()
-    }
+    // Handle worst case ASAP and keep code flat
+    if (!result) throw new IncompleteBufferError()
+
+    buf.consume(result.bytesConsumed)
+    return result.value
   }
 
-  function tryDecode (buf, offset) {
-    offset = offset === undefined ? 0 : offset
-    var bufLength = buf.length - offset
-    if (bufLength <= 0) {
-      return null
-    }
+  function tryDecode (buf, initialOffset = 0) {
+    const bufLength = buf.length - initialOffset
+    if (bufLength <= 0) return null
+
+    var offset = initialOffset
+    const first = buf.readUInt8(offset++)
 
-    var first = buf.readUInt8(offset)
     var length
     var result = 0
-    var type
-    var bytePos
 
-    if (!hasMinBufferSize(first, bufLength)) {
-      return null
+    const size = SIZES[first] || -1
+    if (bufLength < size) return null
+
+    const inRange = (start, end) => (first >= start) && (first <= end)
+
+    // if (inRange(0x00, 0x7f)) TODO positive fixint
+    // if (inRange(0x80, 0x8f)) TODO fixmap
+    // if (inRange(0x90, 0x9f)) TODO fixarr
+    // if (inRange(0xa0, 0xbf)) TODO fixstr
+    if (inRange(0xc0, 0xc3)) return decodeConstants(first)
+    if (inRange(0xc4, 0xc6)) return decodeBuffers(buf, offset, size - 1)
+    if (inRange(0xc7, 0xc9)) return decodeExts(buf, offset, size - 2)
+    if (inRange(0xca, 0xcb)) return decodeFloat(buf, offset, size - 1)
+    if (inRange(0xcc, 0xcf)) return decodeUnsignedInt(buf, offset, size - 1)
+    if (inRange(0xd0, 0xd3)) return decodeSigned(buf, offset, size - 1)
+    if (inRange(0xd4, 0xd8)) return decodeFixExts(buf, offset, size - 2)
+    if (inRange(0xd9, 0xdb)) return decodeStr(buf, offset, size - 1)
+    // if (inRange(0xde, 0xdf)) return decodeMap(...)
+    // if (inRange(0xe0, 0xff)) TODO negative fixint
```

Comment:
I constantly rerunning tests and apply code transformations so that all tests are always passed. So, the answer to question --- no --- functionality is the same. But refactoring is WIP. Map decoding is now implemented the way it has been before --- inside the switch statement. I will factor out to a new function ASAP.